packages/editor/src/bindings/post-meta.js

```diff
@@ -12,6 +12,7 @@ import { store as editorStore } from '../store';
 export default {
 	name: 'core/post-meta',
 	label: _x( 'Post Meta', 'block bindings source' ),
+	usesContext: [ 'postId', 'postType' ],
```

Comment:
Where do we add these today (to block registration). I was expecting some code to be removed in this PR no?

Reply:
We add the context to the block type registration server side, so it will have to be removed there. I'm not sure if it's needed there or also added right in time. My understanding is that it can be adjusted too to not rely on the block type, cc @SantosGuillamot https://github.com/WordPress/gutenberg/pull/58554#discussion_r1484296259

Reply:
We need to do the same for backend sources. Fetch the actual global variable of the context or something like that. then we'd be able to remove the hook. It's too bad that it's not a "named hook" so we can't really unhook it from Gutenberg and we'd need to make the change on Core.

Reply:
I've been taking a look at the server part, and I believe we could add the following logic to the bindings processing [here](https://github.com/WordPress/wordpress-develop/blob/3d139fdf61ae62b51ac26ad28fd2efd89758f173/src/wp-includes/class-wp-block.php#L267).

```php
if ( ! empty( $block_binding_source->uses_context ) ) {
	foreach ( $block_binding_source->uses_context as $context_name ) {
		if ( array_key_exists( $context_name, $this->available_context ) ) {
			$this->context[ $context_name ] = $this->available_context[ $context_name ];
		}
	}
}
```

If we do that, we could get rid of [the old method](https://github.com/WordPress/wordpress-develop/blob/3d139fdf61ae62b51ac26ad28fd2efd89758f173/src/wp-includes/class-wp-block-bindings-registry.php#L192-L204) that changes `uses_context` when the source is registered.

I've tested it, and it seems to work fine. And it only adds the context to the blocks with bindings.

Is that what you had in mind?

Reply:
Yes, that sounds like it.

Reply:
Great. Then we have to make the editor changes in Gutenberg and the server ones in core.

How is this usually handled? Can we merge this PR and start a different one for the server in `core`?

cc: @gziolo

Reply:
We need to support old versions in Gutenberg too, so we need to make the changes in Gutenberg as well for old versions.

Reply:
I’ve been looking at it but I am not sure how to replicate the same in Gutenberg. The problem is that the `available_context` is a [protected property](https://github.com/SantosGuillamot/wordpress-develop/blob/ed1d8460a4a31bb8d0ca25f99010e6a5b3d5ad5d/src/wp-includes/class-wp-block.php#L59), so I am not able to access it in Gutenberg. Unless there is another way I’m not aware of, this means that in Gutenberg, we can only extend `usesContext` during the block registration, which is what we are doing right now in the compatibility functions through the `register_block_type_args` filter: [link](https://github.com/WordPress/gutenberg/blob/ea946fdf9aa39d1504c1027c2ba8b8435e87fc10/lib/compat/wordpress-6.5/block-bindings/class-wp-block-bindings-registry.php#L189-L203).

The difference is that instead of modifying the context while processing the bindings, we extend it when a source is registered, hence extending it even if the block doesn’t use bindings.

Is it okay to make the changes in core and keep this compatibility for old versions?
- From WordPress 6.6: Context is only extended when bindings exist in the block.
- In old versions: Context is extended whenever a source is registered, even if the block doesn’t use bindings.

Any idea how to approach this differently?

Reply:
That sounds like a good plan to me.

Reply:
We considered both approaches in the WordPress 6.5 release cycle. We ended up going with the approach that exposes context for every block that might need it to work with the potentially established connections with block binding sources. It was a simpler approach as we didn't have to apply any changes on the client. The downside is that there is often context exposed even when it's not needed.

As before, I think the revised approach is also viable and presents different challenges. The biggest unknown exists on the server because the whole context for the current block gets computed in the constructor of the `WP_Block`:

https://github.com/WordPress/wordpress-develop/blob/66b5d25be2b5e69833bacfe8f64de660ffc2bfa9/src/wp-includes/class-wp-block.php#L132-L153

It happens inside `render_block`, which is during the rendering phase, so that could have a positive impact on the block registration and editor's bootstrap. On the other hand, it's going to get processed for every individual block that gets rendered on the front end. The key question is if we can have a good strategy for getting the list of registered block sources and their expected context names.

Reply:
This is interesting. I haven't seen this idea before:

https://github.com/WordPress/wordpress-develop/blob/13a63f2730e032967a13eb84081bfc9a9e45bb7a/src/wp-includes/class-wp-block.php#L268-L275

```php
// Add the necessary context defined by the source.
if ( ! empty( $block_binding_source->uses_context ) ) {
	foreach ( $block_binding_source->uses_context as $context_name ) {
		if ( array_key_exists( $context_name, $this->available_context ) ) {
			$this->context[ $context_name ] = $this->available_context[ $context_name ];
		}
	}
}
```

So that would only happen for existing block binding connections locally just before the moment the final values gets computer. That seems very performant.

Reply:
I tried to replicate what this PR does in the server and it's the closest thing I could think of. I also agree that it seems much more performant than the previous implementation.

Reply:
The more I think about it, the more I like it. Great teamwork with the proposed refactoring. It seems like it should be considered as progressive enhancement since the `useContext` is correctly set on block types on the server in WP 6.5, so any changes applied in the Gutenberg plugin won't change much which is great. In WP 6.6, when the block types aren't fed with the same `useContext` from the block binding sources all the time no matter if the context is consumed, then we will seamlessly move that handling to the `WP_Block` class. We might not even need to change anything in the back compat layer in the Gutenberg plugin.